From b0e9703f0cb8fe8c6d26c217b02f0a36dde5f212 Mon Sep 17 00:00:00 2001
From: Christian Hewitt <christianshewitt@gmail.com>
Date: Tue, 31 Mar 2026 07:27:16 +0000
Subject: [PATCH 10/23] media: meson: vdec: fix workspace DMA leak on H.264 SEI
 alloc failure

In codec_h264_start(), if the SEI buffer allocation fails the function
returns -ENOMEM without freeing the workspace buffer that was allocated
just before it. This leaks SIZE_WORKSPACE (0x1ee000) bytes of DMA
coherent memory.

Free the workspace buffer on the SEI allocation error path.

Fixes: 2db3a7bcef41 ("media: meson: vdec: bring up H.264 decoding")
Signed-off-by: Christian Hewitt <christianshewitt@gmail.com>
---
 drivers/staging/media/meson/vdec/codec_h264.c | 6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

diff --git a/drivers/staging/media/meson/vdec/codec_h264.c b/drivers/staging/media/meson/vdec/codec_h264.c
index f0f4a76462df5..ed5669c8612eb 100644
--- a/drivers/staging/media/meson/vdec/codec_h264.c
+++ b/drivers/staging/media/meson/vdec/codec_h264.c
@@ -184,8 +184,12 @@ static int codec_h264_start(struct amvdec_session *sess)
 	/* Allocate some memory for the H.264 SEI dump */
 	h264->sei_vaddr = dma_alloc_coherent(core->dev, SIZE_SEI,
 					     &h264->sei_paddr, GFP_KERNEL);
-	if (!h264->sei_vaddr)
+	if (!h264->sei_vaddr) {
+		dma_free_coherent(core->dev, SIZE_WORKSPACE,
+				  h264->workspace_vaddr, h264->workspace_paddr);
+		h264->workspace_vaddr = NULL;
 		return -ENOMEM;
+	}
 
 	amvdec_write_dos_bits(core, POWER_CTL_VLD, BIT(9) | BIT(6));
 
